route/src/route_generator_worker.cpp

```diff
@@ -460,16 +470,48 @@ namespace route {
                 cte_count_++;
 
                 if(cte_count_ > cte_count_max_) //If the distance exceeds the crosstrack distance a certain number of times, report that the route has been departed
-                    return true;
+                    {
+                        cte_count_ = 0;
+                        return true;
+                    }
+                else 
+                    return false;
+            }
+        else
+            {
+                cte_count_ = 0;
+                return false;
             }
  
     }
 
 
-    void RouteGeneratorWorker::set_CTE_counter(int cte_counter)
+
+    lanelet::ConstLanelet RouteGeneratorWorker::get_closest_lanelet_from_vector_llts (std::vector<std::pair<double, lanelet::ConstLanelet::ConstType>> list_of_pair)
     {
-        cte_count_ = cte_counter;
+        std::vector<std::pair<double, lanelet::ConstLanelet::ConstType>>  filtered;
+        double min = 1000.0;
+        lanelet::ConstLanelet min_llt;
+        for(auto i : list_of_pair)
+        {
+            for(auto j: route_llts)
+            {
+                if(i.second.id() == j.id())
+                    filtered.push_back(i);
+            }
```

Comment:
if we have route_llts, as a set, we would have O(N) operation, while this is O(N^2). We can save route_llt_ids as a set if it is hard to search through lanelets. And later we can query the lanelet itself back from laneletLayer with the ids.

Reply:
Are you talking about a set in terms of some specific detail of the CPP version of a set? Is it backed by a hash-table of some kind? AFAIK a set on it's own would still require iteration in this internal loop to check all elements of the set to see if the ID in question belongs to the set.

Alternatively, is it plausible to restructure this computation in such a way that we only ever get the distance to route lanelets anyway? As I understand it, currently we're using a dijkstra-style method that gets the distance from a single point to all lanelets in our map-graph. Then we check all of those (distance, lanelet) pairs against a list of route_lanelets to determine which are relevant, then find the min among those. Is it possible/more efficient to just do a:

```
min_dist = inf
min_lanelet = null
FOR lanelet IN route:
  dist = get_dist(point, lanelet)
  if (dist < min_dist):
    min_lanelet = lanelet
    min_dist = dist
return min_lanelet
```

style solution? Would be O(N) in terms of our top-level lanelet search. If the underlying distance calculation is still djikstra's then I'm not sure we realize a net gain though, in fact it's probably worse if we run that calculation multiple times.

Reply:
After looking at the documentation, I follow now on the set bit. The common CPP implementation of set lookup is O(log N) though, so overall it'd be O(N log N), I think we could use a hash-set or similar though to get the inner lookup to O(1) though. Probably overkill, I'm not sure how big this data set ends up being at run-time anyway to see these asymptotic performance differences matter.

Reply:
That makes a lot of sense. I was fixated on finding closest lanelets first and then filtering route from there. We could have just iterated through route lanelets and got the closest lanelet as current_lanelet there.